### PR TITLE
Diagnostics: Warnings, Error Context, and Results

### PR DESCRIPTION
**Pull request by cnr**
# Overview

## Diagnostics changes

This PR updates our Diagnostics effect to add a few new primitives related to warnings and adding user-friendly context to errors:

```hs
-- | Emit a warning
warn :: (ToDiagnostic diag, Has Diagnostics sig m) => diag -> m ()

-- | When the provided action fails, emit a warning
warnOnErr :: (ToDiagnostic warn, Has Diagnostics sig m) => warn -> m a -> m a

-- | When the provided action fails, annotate its error with the provided
-- context
errCtx :: (ToDiagnostic ctx, Has Diagnostics sig m) => ctx -> m a -> m a
```

The `ToDiagnostic` typeclass was moved to `Diag.Diagnostic`.

Motivation, examples, and additional documentation can be found in the new
[diagnostics doc][diagnostics_md].

## Result/ResultT

To support the new primitives, Diagnostics uses a new underlying type, [`Result`
from `Diag.Result`](https://github.com/fossas/fossa-cli/blob/diagwarn/src/Diag/Result.hs). The related monad transformer is [`ResultT` from `Diag.Monad`](https://github.com/fossas/fossa-cli/blob/diagwarn/src/Diag/Monad.hs).

## Stack

To reduce the complexity of `Diagnostics` and `ResultT`, the `context` operation
was migrated into its own effect, `Control.Effect.Stack`.

`Stack` additionally gains two new operations:

```hs
-- | Get the current scope stack
getStack :: Has Stack sig m => m [Text]

-- | Run the provided action with an empty callstack
withEmptyStack :: Has Stack sig m => m a -> m a
```

..which additionally reduces the complexity of our hack to "display task
progress" in the form of `stickyLogStack` (renamed from `stickyDiag`)

As an added benefit, our debug bundle code similarly drops some complexity.

## Error/warning display changes

This PR changes both _when_ and _how_ we display errors and warnings.

Previously, when a discovery/analysis task **succeeds**, that task would emit
none of its "warnings" (which previously were limited to "errors we `recover`ed
from"). In this PR, a successful task prints the warnings created with `warn`
and `warnOnErr`.

Previously, when a discovery/analysis task **fails**, that task would emit all
of the errors it `recover`ed from. We maintain this behavior. Additionally, we
print the warnings created with `warn` and `warnOnErr`.

### Setuptools

In the setuptools analyzer, we previously silently ignored parse failures for
`setup.py` and `*req*.txt` files. After this PR, we are much much noisier about
parse failures.

Because our `setup.py` parser is so brittle, customers with setuptools projects
are **almost always** going to see a warnings for `setup.py` parse failures.

## Remaining work

- [ ] The display format for failure/success of discovery and analysis tasks
      needs to be finalized

## Future work

- I've left a few `TODO(warnings)` comments; these will need to be addressed
  in future PRs

## Acceptance criteria

- The `Diagnostics` effect fits our needs for reporting errors and warnings
  reporting (see [diagnostics.md][diagnostics_md])

## Testing plan

- Lots and lots of manual testing
  - Analyzing local projects, inducing intentional failures for various steps in
    analyzers (e.g., running with `mvn` in scope, running again without `mvn` in
    scope), and ensuring that results are identical pre- and post-PR
  - Ensuring debug bundles continue to work identically pre- and post-PR
- These changes were propagated to our unit and integration tests, which continue to pass
- Tests to validate the semantics of diagnostics constructions

## Risks

Diagnostics is our most-often-used effect in our codebase, and this entirely
replaces the datatype used by the `DiagnosticsC` carrier.

If implemented incorrectly, this could result in dropped warnings or errors.

I've taken care to break down the functionality of the underlying types,
`Result` and `ResultT`, very granularly to make testing for their behavior and
semantics very straightforward.

## References

- Addresses https://github.com/fossas/team-analysis/issues/852

- Partially addresses https://github.com/fossas/team-analysis/issues/853

## Checklist

- [ ] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.

[diagnostics_md]: https://github.com/fossas/fossa-cli/blob/diagwarn/docs/contributing/diagnostics.md

### CODE REVIEW

File: src/Control/Carrier/Stack.hs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import Control.Carrier.Reader
+import Control.Effect.Stack as X
+import Control.Monad.IO.Class (MonadIO)
+import Control.Monad.Trans (MonadTrans, lift)
+import Data.Text (Text)
+
+newtype StackC m a = StackC {runStackC :: ReaderC [Text] m a}
+  deriving (Functor, Applicative, Monad, MonadIO)
+
+runStack :: [Text] -> StackC m a -> m a
```
Comment (meghfossa) [inline src/Control/Carrier/Stack.hs:22]: [nit] we have lots of `runStack []`, prefer if we collapse the default value in signature.

Prefer:
```haskell
runStack :: StackC m a -> m a
runStack = runStack' []

runStack' :: [Text] -> StackC m a -> m a
runStack' initial = runReader initial . runStackC
```

**Review by scruffystuffs — COMMENTED** [review]
Looks good overall, I'd like to see rendering examples before I approve.  The `Result/ResultT/Stack/Diag` split is solid.

Left a few nits, but I didn't see any major issues yet.

File: src/Control/Carrier/Stack.hs (hunk 1 above)
Comment (cnr) [reply src/Control/Carrier/Stack.hs]: Addressed in https://github.com/fossas/fossa-cli/commit/fa2a7fd9b768bee9a068e55f77545e5683ba9107

File: src/Control/Carrier/Stack.hs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+runStack initial = runReader initial . runStackC
```
[the remark this answers is not in the document]
Comment (cnr) [reply src/Control/Carrier/Stack.hs]: Addressed in https://github.com/fossas/fossa-cli/commit/fa2a7fd9b768bee9a068e55f77545e5683ba9107

File: docs/contributing/diagnostics.md [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+example = do
+  ...
+  fatal MyErrType
+
+data MyErrType
+instance ToDiagnostic MyErrType
+```
+
+#### Adding error context to an action
```
[the remark this answers is not in the document]
Comment (cnr) [reply docs/contributing/diagnostics.md:140]: Addressed in https://github.com/fossas/fossa-cli/commit/524c2f1ef09b93369e4e914e086fe07bba7e0916 (along with an explanation for the intended use of `withWarn`)

File: src/Control/Carrier/Diagnostics.hs [hunk 4]
```diff
@@ -20,93 +18,119 @@ module Control.Carrier.Diagnostics (
...
+-- - On success, the associated warnings are logged with the provided
+--   @sevOnSuccess@ severity
+withResult :: Has Logger sig m => Severity -> Severity -> Result a -> (a -> m ()) -> m ()
+withResult _ sevOnSuccess (Success ws res) f = do
+  case renderSuccess ws of
+    Nothing -> pure ()
+    Just rendered -> Effect.Logger.log sevOnSuccess rendered
+
+  f res
+withResult sevOnErr _ (Failure ws eg) _ = Effect.Logger.log sevOnErr (renderFailure ws eg)
```
[the remark this answers is not in the document]
Comment (cnr) [reply src/Control/Carrier/Diagnostics.hs]: Addressed in https://github.com/fossas/fossa-cli/commit/3f64060841a287c840332a1eec16fc51015f4c83

File: src/Control/Carrier/Diagnostics.hs [hunk 5]
```diff
@@ -20,93 +18,119 @@ module Control.Carrier.Diagnostics (
...
+      case renderSuccess ws of
+        Nothing -> pure ()
+        Just rendered -> logWarn rendered
+
+      pure (Just a)
+
+-- result <- runDiagnosticsIO diag
+-- case result of
+--   Left failure -> logErrorBundle failure >> pure Nothing
+--   Right success -> pure $ Just success
```
[the remark this answers is not in the document]
Comment (cnr) [reply src/Control/Carrier/Diagnostics.hs]: Addressed in https://github.com/fossas/fossa-cli/commit/3b8742e01a1c67b3ff21df0ec4f70af32b6916ab

File: src/Control/Carrier/Diagnostics.hs [hunk 6]
```diff
@@ -20,93 +18,119 @@ module Control.Carrier.Diagnostics (
...
+-- "invert" a Maybe into a functorial context
+distributeMaybe :: Functor ctx => ctx () -> Maybe (ctx a) -> ctx (Maybe a)
+distributeMaybe _ (Just ctx) = Just <$> ctx
+distributeMaybe ctx Nothing = Nothing <$ ctx
+{-# INLINE distributeMaybe #-}
+
+-- "distribute" a Result into a DiagnosticsC computation
+hdlResult :: Applicative m => Result (DiagnosticsC m a) -> m (Result a)
+hdlResult (Failure ws eg) = pure (Failure ws eg)
+hdlResult (Success ws m) = addWarns <$> ResultT.runResultT (runDiagnosticsC m)
```
[the remark this answers is not in the document]
Comment (cnr) [reply src/Control/Carrier/Diagnostics.hs]: Addressed in https://github.com/fossas/fossa-cli/commit/aad2f2378940a171b1d9a3cd873efe5617acb8c2

**Review by meghfossa — APPROVED** [review]
LGTM.

I performed following manual tests:

✅ Without network connection.
✅  Running elixir strategy without compiled project (`warn`)
✅  Pub alternative strategy. 
✅  Setup.py parser error.

For `WarnErr` I did not perform manual tests - I see that they are only in Gradle/Scala Project Discovery.

Only one caveat I have with the current format (as is) is when we have an error emitted as part of warning - final output is a bit counter intuitive. We at top level say, analysis was finished with warnings, but we have an error headings. 

I know this format is expected to change with future work, but if since this is merging on to main, prefer if we get thumbs up from rest of the team (that we will have not have ideal stdout format for a bit).

```
[ WARN] ----------
  A task succeeded with warnings

  Warning 1

    Failed to parse python file

    >>> Relevant errors

      Error 1

        Error parsing file /Users/megh/Work/upstream/example-projects/python/setuptools/setup.py:
            /Users/megh/Work/upstream/example-projects/python/setuptools/setup.py:1:1:
              |
            1 | <empty line>
              | ^
            unexpected end of input
            expecting "install_requires"


        Traceback:
          - Parsing file '/Users/megh/Work/upstream/example-projects/python/setuptools/setup.py'
          - Analyzing setup.py
          - Setuptools
          - Project Analysis
```

File: src/Control/Carrier/Stack/StickyLog.hs [hunk 7]
```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Carrier.Stack.StickyLog (
+  StickyLogStackC (..),
+  stickyLogStack,
+) where
+
+import Console.Sticky qualified as Sticky
+import Control.Carrier.Reader
```
[the remark this answers is not in the document]
Comment (cnr) [reply src/Control/Carrier/Stack/StickyLog.hs]: Addressed in https://github.com/fossas/fossa-cli/commit/ea2170b104d9bf80c161037e1241771c7d3d983e

**Comment by cnr** [comment]
I've added a [temporary module, `DiagExamples`](https://github.com/fossas/fossa-cli/blob/diagwarn/src/DiagExamples.hs), that helps to illustrate how errors/warnings are rendered for common scenarios

The examples are simplified, and the error/warning message contents are obviously terrible, but the structure is what matters here.

It can be loaded and run with:

```
cabal repl
>>> :l DiagExamples
>>> examples
```

---

I've also attached the rendered examples below:

image

image

image

image

**Pull request merged**